### PR TITLE
GH-4131 Introduce custom AggregateFunctionCall abstraction

### PR DESCRIPTION
**Pull request by tokovach**
GitHub issue resolved: #4131

Briefly describe the changes proposed in this PR:

Introduced custom AggregateFunctionCall abstraction that allows custom aggregate functions to be introduced through CustomAggregateFunctionRegistry. 

Updated sparql.jjt compile file and subsequently SyntaxTreeBuilder to allow distinct to be parsed for custom aggregate function calls.

Moved around abstract class for aggregate functions (Aggregate and renamed it to AggregateFunction) as well as the AggregateCollector in order to be able to define custom aggregate functions (whose signature can be the same as a standard function call) through a custom function registry (where they are validated inside the TupleExprBuilder).

### CODE REVIEW

**Comment by tokovach** [comment]
Will add unit & functional tests covering custom AggregateCustomFunction execution.

File: core/queryparser/sparql/src/main/java/org/eclipse/rdf4j/query/parser/sparql/ast/JJTSyntaxTreeBuilderState.java [hunk 1]
```diff
@@ -5,10 +5,8 @@
  * are made available under the terms of the Eclipse Distribution License v1.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/org/documents/edl-v10.php.
- *
- * SPDX-License-Identifier: BSD-3-Clause
- *******************************************************************************/
-/* Generated By:JavaCC: Do not edit this line. JJTSyntaxTreeBuilderState.java Version 7.0.11 */
+ ******************************************************************************/
+/* Generated By:JavaCC: Do not edit this line. JJTSyntaxTreeBuilderState.java Version 5.0 */
```
Comment (jeenbroekstra) [inline core/queryparser/sparql/src/main/java/org/eclipse/rdf4j/query/parser/sparql/ast/JJTSyntaxTreeBuilderState.java:9]: Looks like you used an older version of JavaCC than the one we most recently upgraded to - worth rerunning with the newer version?

**Review by jeenbroekstra — COMMENTED** [review]
I've only given this a brief first glance over, but looks really solid, nice work!

File: core/queryparser/sparql/src/main/java/org/eclipse/rdf4j/query/parser/sparql/ast/JJTSyntaxTreeBuilderState.java (hunk 1 above)
Comment (tokovach) [reply core/queryparser/sparql/src/main/java/org/eclipse/rdf4j/query/parser/sparql/ast/JJTSyntaxTreeBuilderState.java:9]: Great, I will rerun it with a newer version.

**Comment by hmottestad** [comment]
## Before
```
```

## After
```
Benchmark                                        Mode  Cnt   Score   Error  Units
SPARQLParseBenchmark.complexPathExpressionQuery  avgt    5  21.765 ± 1.686  ms/op
SPARQLParseBenchmark.selectValuesQuery           avgt    5  20.218 ± 0.265  ms/op
```

**Comment by hmottestad** [comment]
I'm going to follow up on the performance regression in a separate PR.

**Comment by hmottestad** [comment]
Just realised that I merged this into main instead of develop :(

**Pull request merged**